### PR TITLE
Add get model

### PR DESCRIPTION
**Pull request by SimonRichardson**
We need the model data for the environ tracker worker. I know we're potentially renaming model_metadata to model, but for now, I need to get a model.

We already have all the information, so I'm just getting the information back out. As part of that, I needed to have a core Model type. This meant moving the existing `core/Model` to `api/types`. The API client package shouldn't be using core model types (we'll be changing this as we go, this is the first step!). Then we can expose a core model type. All the churn then becomes, you're not allowed to bring domain into core, the credential type is in the wrong place. Moving core to domain will solve this, but I'm not going to tackle that in this PR.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Just bootstrapping should work, as I'm just moving things around and exposing one method.

```sh
$ juju bootstrap lxd test
```

## Links

**Jira card:** JUJU-5381

### CODE REVIEW

File: domain/credential/types.go [hunk 1]
```diff
@@ -3,79 +3,6 @@
...
 
-import (
-	"fmt"
-
-	"github.com/juju/errors"
-	"github.com/juju/names/v5"
-)
-
-// ID represents the id of a cloud credential.
-type ID struct {
```
Comment (SimonRichardson) [inline domain/credential/types.go:14]: Core can import domain, so we need to lift this out of domain.

File: api/types/types.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+// ModelType indicates a model type.
+type ModelType string
+
+const (
+	// IAAS is the type for IAAS models.
+	IAAS ModelType = "iaas"
+
+	// CAAS is the type for CAAS models.
+	CAAS ModelType = "caas"
```
Comment (tlm) [inline api/types/types.go:14]: Should we introduce a test to make sure these values haven't drifted from core? I know it probably isn't needed but I like things to go bang if something changes.

File: apiserver/common/credentialcommon/cloudcredential.go [hunk 3]
```diff
@@ -12,7 +12,7 @@ import (
 	apiservererrors "github.com/juju/juju/apiserver/errors"
 	"github.com/juju/juju/apiserver/facade"
 	"github.com/juju/juju/cloud"
-	"github.com/juju/juju/domain/credential"
+	"github.com/juju/juju/core/credential"
```
Comment (tlm) [inline apiserver/common/credentialcommon/cloudcredential.go:15]: This is the best change. Thank you so much.

File: core/model/model.go [hunk 4]
```diff
@@ -49,6 +42,40 @@ func (m ModelType) IsValid() bool {
...
+	Cloud string
+
+	// CloudRegion is the region that the model will use in the cloud.
+	CloudRegion string
+
+	// Credential is the id attributes for the credential to be associated with
+	// model. Credential must be for the same cloud as that of the model.
+	// Credential can be the zero value of the struct to not have a credential
+	// associated with the model.
+	Credential credential.ID
```
Comment (tlm) [inline core/model/model.go:70]: Doesn't need to be changed as part of this PR. More of a question for you in general but I am thinking this needs to get swapped over to credential id (uuid) at some stage. Get us off this natural key as early as possible.

File: domain/credential/types.go (hunk 1 above)
Comment (tlm) [reply domain/credential/types.go:14]: THANK YOU!!!

**Review by tlm — APPROVED** [review]
Looks good. Send it!

File: domain/model/service/service.go [hunk 5]
```diff
@@ -133,6 +136,17 @@ func (s *Service) CreateModel(
 	return uuid, s.st.Create(ctx, uuid, args)
 }
 
+// Model returns the model associated with the provided uuid.
+// The following error types can be expected to be returned:
+// - [modelerrors.ModelNotFound]: When the model does not exist.
+func (s *Service) Model(ctx context.Context, uuid coremodel.UUID) (coremodel.Model, error) {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply domain/model/service/service.go:142]: I'm not sure anymore, I thought it was `Model`, `SaveModel`, `DeleteModel`, as this is more idiomatic.

**Pull request merged**